src/aws/flb_aws_util.c

```diff
@@ -927,3 +933,84 @@ flb_sds_t flb_get_s3_key(const char *format, time_t time, const char *tag,
         }
         return NULL;
 }
+
+/*
+ * Count the substring number in string.
+ */
+int count_substring(char *str1, char *str2){
+    int x = 0;
+    char *p;
+
+    do {
+        p = strstr(str1, str2);
+        if (p != NULL) {
+            str1 = p + 1;
+            x = x + 1;
+        }
+    }
+    while(p != NULL);
+    return x;
+}
+
+/*
+ * Adjust the time_key output format according to time_key_format.
+ */
+int create_time_format_string(char **out_buff, size_t *out_size,
+                              char *time_key_format, const struct tm *timeptr,
+                              size_t len, struct flb_time *tms)
+{
+    char millisecond_string[FLB_MILLISECOND_STRING_LENGTH+1];
+    char nanosecond_string[FLB_NANOSECOND_STRING_LENGTH+1];
+    char *tmp_buff_time_str;
+    char *mem_for_time_key_format;
+    char *tmp_out_buff;
+    size_t tmp_out_size;
+
+    /* guess the maximum length needed for out_buff, replace %F to 3 chars and %L to 9 chars */
+    len = 12*strlen(time_key_format)+count_substring(time_key_format, FLB_MILLISECOND_STRING)+7*count_substring(time_key_format, FLB_NANOSECOND_STRING);
+    mem_for_time_key_format = (char *)calloc(strlen(time_key_format), sizeof(char));
```

Comment:
https://linux.die.net/man/3/calloc

Read that. On most machines, this allocs 4 bytes for a char in strlen(time_key_format) blocks. Usually we alloc in one block. Also we must use `flb_calloc`